### DISCUSSION TITLE
Enable a decent way to build / test Meteor in CI

### DISCUSSION THREAD

**Discussion by satyavh** [meteor/meteor#11519] [Feature Requests]
The build process needs to happen locally, which is really inconvenient and unprofessional when working with Meteor in teams / professionally. There are a couple of attempts by people (such as Disney) to provide a Meteor docker container, which allows you to build Meteor in CI - and if you persist you can make it work. You will run into issues such as Galaxy only accepting secrets from settings.json, or how to test your code before building. I made it happen but I can't recommend the journey.

Imho for Meteor to grow up, this needs to be easily possible.

**Comment by StorytellerCZ** [comment]
So what do you propose should be done? If we could split this into concrete steps that would be a great step towards making this happen.

**Comment by filipenevola** [comment]
Hi @satyavh we agree that easier deploys are going to help in our growth.

We are working at the moment in this feature on Meteor Cloud.

The idea is simple, you create a new app on Galaxy (or select an existent one) and then the branch from a GitHub repository that you have access to.

After that new commits to this branch are going to create a new deploy on Galaxy.

Are there other use cases that you believe are not going to be covered by this new feature that is coming to Meteor Cloud?

---

You also talk about testing, what is the challenge for you in this area?

Usually I use Jenkins on my projects to run my tests without any issues so I would like to understand more the issues that you are facing.

**Reply by satyavh** [reply]
Git-based deployments would be a huge step forward.

But for testing in CI pipelines (for example circleci/gitlab/etc), Meteor needs to be build in the pipeline in order to run tests prior to deployment. That means it needs:

1.  a docker image
2. then install meteor in the container
3. build code
4. setting env
5. run tests, etc.

The docker image is the main problem. Best would be if Meteor would provide and maintain docker image that already has specific Meteor versions preinstalled. There are community attempts, the best example by [Disney](https://github.com/disney/meteor-base). Check the code for its relative complexity.

**Reply by filipenevola** [reply]
What problems are you having with Docker?

I was playing around last week with Docker and I was able to use this image to run a Meteor app.

```Dockerfile
FROM node:14.17-buster

ENV METEOR_ALLOW_SUPERUSER=true
ENV ROOT_URL="http://localhost:3000"

RUN curl https://install.meteor.com/ | sh

WORKDIR /code

COPY package.json package.json
COPY package-lock.json package-lock.json

COPY .meteor/release .meteor/release
COPY .meteor/platforms .meteor/platforms
COPY .meteor/packages .meteor/packages
COPY .meteor/versions .meteor/versions

RUN meteor npm install
RUN meteor --get-ready

COPY . .

RUN chmod -R 700 /code/.meteor/local

EXPOSE 3000

CMD ["meteor", "--settings", "/code/private/env/production/settings.json", "--exclude-archs", "web.browser.legacy" ]
```

The only thing that I'm not happy in this Dockerfile is this `ENV METEOR_ALLOW_SUPERUSER=true`.

Besides that what other pieces do you think are missing?

The same way that I'm running meteor as a regular app you could change this last line to run meteor tests.

**Comment by StorytellerCZ** [comment]
Just encountered this package for mocking: https://mswjs.io/
I think if we can create an extension for Meteor it could be super helpful or to have something build-in.
